### ISSUE TITLE
Schema qualifier for enums

### ISSUE THREAD

**Issue by parm-rm** [tgriesser/knex#3307]
# Environment

Knex version: 0.17.6
Database + version: PostgreSQL 11.4
OS: Manjaro Linux 18.0.4

# Feature discussion / request

While creating a table with an explicit schema and I'm looking to use a native enum as one of the columns. I haven't found any documentation about how to qualify an enum declaration with schema name.

I've tried the following
```
table.enum('bazz_type', [ 'foo', 'bar' ], {
	useNative : true,
	enumName  : 'other_schema.bazz_types',
});
```
but it lead to creation of `other_schema.bazz_types` under `public` schema
```
Debugging output:
'create type ' +
'"other_schema.bazz_types" ' +
"as enum ('foo', 'bar')
```
One possible solution could be to add a `schema` string option under `EnumOptions`.

**Comment by elhigu** [comment]
Yep, looks like a bug that enum type creation does not respect withSchema.

https://runkit.com/embed/e0rde4cgbloq

for a workaround this should work though:

```
knex.schema.withSchema('foo').table('table', t => {
  t.enum('enum', ['foo', 'bar'], { useNative: true, enumName: knex.ref('foo.enu_type') });
})```

**Comment by elhigu** [comment]
Do you want table and enum type to live in different schemas?

**Comment by parm-rm** [comment]
> Do you want table and enum type to live in different schemas?

No, just wanted it to respect table schema context. Having an schema override for enums too is "nice-to-have".

**Comment by elhigu** [comment]
Ok, I'll label this as a bug then :)

**Comment by parm-rm** [comment]
> Yep, looks like a bug that enum type creation does not respect withSchema.
> 
> https://runkit.com/embed/e0rde4cgbloq
> 
> for a workaround this should work though:
> 
> ```
> knex.schema.withSchema('foo').table('table', t => {
>   t.enum('enum', ['foo', 'bar'], { useNative: true, enumName: knex.ref('foo.enu_type') });
> })
> ```

That workaround doesn't seem to work. Instead of generating the following SQL
```
create type "foo"."enu_type" as enum ('foo', 'bar')
```
it generates
```
create type ""foo"."enu_type"" as enum ('foo', 'bar')
```
which is invalid and leads to `zero-length delimited identifier at or near """"` error.

**Comment by elhigu** [comment]
Right I missed that... so looks like there is no nice workaround for that, but one must use knex.schema.raw() to create type manually... That double quoting problem seems to be also when type name is used when creating enum column :(

**Comment by mercmobily** [comment]
Hello! Thanks for reporting this issue and providing a clear use case.


This issue has been resolved in version 0.19.3 (released August 2019).


The problem you identified was indeed a bug. When using `withSchema()` with native enums, the enum type was being created in the public schema rather than respecting the table's schema context. This happened because the code in `lib/dialects/postgres/schema/pg-columncompiler.js:35` wasn't checking `this.tableCompiler.schemaNameRaw` when constructing the enum type name.


Two pull requests fixed this issue completely:

PR #3400 added support for automatically respecting the table schema when using `withSchema()`. The fix modified the `enu()` method to check `this.tableCompiler.schemaNameRaw` and prepend it to the enum type name when present. This ensures that when you use `withSchema('foo')`, the enum type is created as `"foo"."enu_type"` rather than just `"enu_type"` in the public schema.

PR #3413 extended this functionality to allow manual schema override via a `schemaName` option in the enum options. This gives you the flexibility to create the enum type in a different schema from the table itself.


The current implementation now supports both use cases:

1. Automatic schema inheritance from `withSchema()`:
```js
knex.schema.withSchema('foo').table('table', t => {
  t.enum('enum', ['foo', 'bar'], { useNative: true, enumName: 'enu_type' });
});
// Creates: create type "foo"."enu_type" as enum ('foo', 'bar')
```

2. Manual schema override using `schemaName`:
```js
knex.schema.withSchema('table_schema').table('table', t => {
  t.enum('enum', ['foo', 'bar'], {
    useNative: true,
    enumName: 'enu_type',
    schemaName: 'type_schema'
  });
});
// Creates: create type "type_schema"."enu_type" as enum ('foo', 'bar')
```

The workaround suggested by @elhigu using `knex.ref('foo.enu_type')` was a good attempt, but as you discovered, it resulted in invalid SQL with double quoting. This is no longer necessary with the fix in place.


I've tested the current codebase and confirmed both scenarios work correctly. The enum types are now properly created in the specified schema, and comprehensive test coverage was added in both unit and integration tests to prevent regression.


**Related issues:**

- #394 (CLOSED): Initial native enum support
- #1699 (OPEN): Updating enum column types in migrations
- #4936 (CLOSED): Invalid query issues with enums on PostgreSQL 13
- #4563 (OPEN): Migrations table created in multiple schemas

**Comment by mercmobily** [comment]
This issue was resolved in version 0.19.3. Please upgrade to the latest version of Knex to get this fix.
